### PR TITLE
Fix useData() hooks and add testing.

### PR DESCRIPTION
**Pull request by mikelehen**
The existing useData() hooks didn't gracefully handle passing the same metrics / regions on every render if the actual array instances changed, e.g.: This would cause an infinite re-render:

```
catalog.useDataForRegionsAndMetrics([region], [metric])
```

Because the array instances would be differente every time.  The core fix is to add a `useCachedArrayIfEqual()` helper that preserves the original array reference if the contents are the same.

This PR also:
* Adds a test to validate the hook behaves correctly now.
* Enables `plugin:react-hooks/recommended` eslint checks.

### CODE REVIEW

File: packages/metrics/src/MetricCatalog/MetricCatalog.ts [hunk 1]
```diff
@@ -255,27 +247,19 @@ export class MetricCatalog {
...
-        })
-        .catch((error) => {
-          console.error(`Error fetching metric data: ${error}`);
-          return { error };
-        });
-    }, [metricList, regions]);
-    return { data };
+    // ESLint complains about calling hooks from classes (React Hook
+    // "useDataForRegionsAndMetrics"cannot be called in a class component.) but
+    // this isn't actually a class component.  It's just a class.
```
Comment (mikelehen) [inline packages/metrics/src/MetricCatalog/MetricCatalog.ts:255]: TBH I'm not 100% sure if it's "okay" to have a hook defined inside a class like I'm doing.  Like maybe having components call `catalog.useData(...)` is going to cause issues?  The alternative would be to export the `useDataForRegionsAndMetrics()` function that this method calls and make everybody use that.  But I prefer it being a class method if it'll work.

File: .eslintrc.json [hunk 2]
```diff
@@ -7,6 +7,7 @@
   "extends": [
     "eslint:recommended",
     "plugin:react/recommended",
+    "plugin:react-hooks/recommended",
```
Comment (mikelehen) [inline .eslintrc.json:10]: This enables the same hook-related lint checks that `create-react-app` apps have by default, including making sure that your `useEffect()` calls specify all of their dependencies, etc.

File: packages/metrics/src/MetricCatalog/MetricCatalog.ts [hunk 3]
```diff
@@ -194,20 +206,13 @@ export class MetricCatalog {
...
-          setData(data);
-        })
-        .catch((error) => {
-          console.error(
-            `Error fetching metric data for ${metric} for ${region.regionId}: ${error}`
-          );
-          return { error };
-        });
-    }, [region, metric]);
-    return { data };
```
Comment (mikelehen) [inline packages/metrics/src/MetricCatalog/MetricCatalog.ts:210]: I realized that these useData() methods can call each other, so the implementation guts only live inside `useDataForRegionsAndMetrics()`

File: packages/metrics/src/MetricCatalog/MetricCatalog.ts [hunk 4]
```diff
@@ -287,3 +271,59 @@ export class MetricCatalog {
...
+  const [data, setData] = useState<MultiRegionMultiMetricDataStore>();
+  let resolvedMetrics = metrics.map((m) =>
+    typeof m === "string" ? catalog.getMetric(m) : m
+  );
+
+  // In order to allow people to pass in a new array of regions / metrics that
+  // contain the same regions / metrics as before without triggering additional
+  // fetches, we need this caching trick.
+  resolvedMetrics = useCachedArrayIfEqual(resolvedMetrics);
+  regions = useCachedArrayIfEqual(regions);
```
Comment (mikelehen) [inline packages/metrics/src/MetricCatalog/MetricCatalog.ts:290]: While I like this approach better than the `JSON.stringify` or string concatenation approaches, it is a bit unconventional so I ended up writing more comments than code, trying to make it clearer.  :-/ Let me know if you have any suggestions / better ideas!

File: packages/metrics/src/MetricCatalog/MetricCatalog.ts [hunk 5]
```diff
@@ -287,3 +271,59 @@ export class MetricCatalog {
...
+ * @param value New array to check against the cached array.
+ * @returns The cached array if its contents is exactly equal to the new value, else the new value.
+ */
+function useCachedArrayIfEqual<T>(value: Array<T>): Array<T> {
+  const ref = useRef(value);
+  if (!isEqual(ref.current, value)) {
+    ref.current = value;
+  }
+  return ref.current;
+}
```
Comment (pnavarrc) [inline packages/metrics/src/MetricCatalog/MetricCatalog.ts:329]: Consider adding a few tests for this function so we can have a more explicit demonstration of how it works

**Review by pnavarrc — APPROVED** [review]
Looks good!

File: packages/metrics/src/MetricCatalog/MetricCatalog.ts (hunk 5 above)
Comment (mikelehen) [reply packages/metrics/src/MetricCatalog/MetricCatalog.ts:329]: Sure, sounds good!

**Pull request merged**